### PR TITLE
[CL-106] use CL's DialogService in Desktop & Browser

### PR DESCRIPTION
**Pull request by willmartian**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

Relies on: https://github.com/bitwarden/clients/pull/5890

## Objective

- Desktop and Browser will now use the CL's `DialogService`
  - The `DialogServiceAbstraction` was deleted, and all imports of it were updated to the CL's `DialogService`.
  - `DialogService.open` is now available in Desktop and Browser, meaning the `ModalService` can be fully deprecated.
  - The `BrowserDialogService` and `ElectronDialogService` were deleted. SimpleDialogs in all clients will now use the CL implementation.
- Resolve circular dependency between `libs/angular` and `libs/components`
  - `SimpleDialogComponent` and related files were moved to `libs/components`.
  - `libs/angular` is now a restricted import in the CL
- `SimpleDialogType` and `SimpleDialogCloseType` are now types instead of enum
  - After moving the enums to the CL, I was running into an issue where they were undefined at runtime in test files. This looks like an issue with `ts-jest` and re-exporting enums in multiple files: https://github.com/kulshekhar/ts-jest/issues/3397
   -  (I also dislike TS enums in general, and since I am diffing all of these files anyways, might as well use this as my chance to get rid of the enum. 😈)

## Code changes

- `.eslintrc.json`: Added restricted imports
- `libs/angular/src/services/modal.service.ts`: Added deprecation note
- `libs/components/src/dialog/dialog.service.ts`: Updated imports; removed DialogServiceAbstraction; added `translate` method that is called by `SimpleDialogComponent`
- `libs/components/src/dialog/simple-dialog/types.ts`: Colocated `SimpleDialog` types
- `libs/shared/tsconfig.libs.json`: Added `@bitwarden/components` import alias
- `libs/angular/src/services/dialog/*`: deleted
- **Everything else**: Updated imports to `DialogServiceAbstraction`; removed `SimpleDialog` enums

## Screenshots

`DialogService.open` with `bit-dialog` in browser extension:

https://github.com/bitwarden/clients/assets/17113462/00cbdc73-f2be-4a9c-9e51-88e8018b546f

`DialogService.open` with `bit-dialog` on desktop:

https://github.com/bitwarden/clients/assets/17113462/774cb044-951f-4d86-8b56-073b1256785e

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -143,7 +143,10 @@
     {
       "files": ["libs/components/src/**/*.ts"],
       "rules": {
-        "no-restricted-imports": ["error", { "patterns": ["@bitwarden/components/*", "src/**/*"] }]
+        "no-restricted-imports": [
+          "error",
+          { "patterns": ["@bitwarden/components/*", "src/**/*", "@bitwarden/angular/*"] }
```
[the remark this answers is not in the document]
Comment (willmartian) [reply .eslintrc.json:148]: Yes. I think we could probably restrict all `@bitwarden` deps in the CL.

File: apps/web/src/app/billing/settings/payment-method.component.ts [hunk 2]
```diff
@@ -100,7 +100,7 @@ export class PaymentMethodComponent implements OnInit {
         content: { key: "cannotPerformInAppPurchase" },
         acceptButtonText: { key: "ok" },
         cancelButtonText: null,
-        type: SimpleDialogType.WARNING,
```
[the remark this answers is not in the document]
Comment (willmartian) [reply apps/web/src/app/billing/settings/payment-method.component.ts:103]: No type safety is lost. String literal unions provide the same type safety as enums without requiring another import. See more info here: https://www.bam.tech/en/article/should-you-use-enums-or-union-types-in-typescript

In `openSimpleDialog`, `type` expects a value of `SimpleDialogType`.

`type SimpleDialogType = "primary" | "success" | "info" | "warning" | "danger";`

If a dev were to pass an unknown string to this field, such as `"foo"`, there would be an editor/compile error.

File: apps/desktop/src/app/tools/export/export.component.ts [hunk 3]
```diff
@@ -67,14 +67,14 @@ export class ExportComponent extends BaseExportComponent implements OnInit {
           os.EOL +
```
[the remark this answers is not in the document]
Comment (willmartian) [reply apps/desktop/src/app/tools/export/export.component.ts]: Resolved in [dc725c8](https://github.com/bitwarden/clients/pull/5875/commits/dc725c8c75521588062b696ad870dd84f1f1b9e5). CC @bitwarden/team-tools-dev

File: libs/components/src/dialog/simple-dialog/types.ts [hunk 4]
```diff
@@ -49,3 +51,7 @@ export type SimpleDialogOptions = {
   /** Whether or not the user can use escape or clicking the backdrop to close the dialog */
   disableClose?: boolean;
 };
+
+export type SimpleDialogCloseType = "accept" | "cancel";
```
[the remark this answers is not in the document]
Comment (willmartian) [reply libs/components/src/dialog/simple-dialog/types.ts]: Yes, resolved in [86b4392](https://github.com/bitwarden/clients/pull/5875/commits/86b43924d283181299d1081698c7d3914213761d).

File: libs/components/src/dialog/dialog.service.ts [hunk 5]
```diff
@@ -88,12 +86,12 @@ export class DialogService extends Dialog implements OnDestroy, DialogServiceAbs
    * @returns `boolean` - True if the user accepted the dialog, false otherwise.
    */
   async openSimpleDialog(simpleDialogOptions: SimpleDialogOptions): Promise<boolean> {
-    const dialogRef = this.open(SimpleConfigurableDialogComponent, {
+    const dialogRef = this.open<SimpleDialogCloseType>(SimpleConfigurableDialogComponent, {
```
Comment (Hinton) [inline libs/components/src/dialog/dialog.service.ts:89]: Should we replace `SimpleDialogCloseType` with a simple boolean?

**Review by Hinton — DISMISSED** [review]

**Review by djsmith85 — DISMISSED** [review]
Changes are looking great! Nice work on this 🚀

File: libs/components/src/dialog/dialog.service.ts (hunk 5 above)
Comment (willmartian) [reply libs/components/src/dialog/dialog.service.ts]: Resolved in [b9db815](https://github.com/bitwarden/clients/pull/5875/commits/b9db815481d323f4aae2bfe42be710efe0305546).

**Review by djsmith85 — APPROVED** [review]

**Review by shane-melton — APPROVED** [review]
Changes related to AC look good!

**Review by Hinton — APPROVED** [review]

**Review by jlf0dev — APPROVED** [review]
Looks good! I feel the same about enums so nice to see the unions. Not importing makes it slightly more difficult to refactor/rename, but seems like everyone else is good with that so 👍

File: apps/web/src/app/billing/settings/payment-method.component.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply apps/web/src/app/billing/settings/payment-method.component.ts:103]: Ty for this useful link: https://www.bam.tech/en/article/should-you-use-enums-or-union-types-in-typescript
TIL!

File: libs/angular/src/services/modal.service.ts [hunk 6]
```diff
@@ -21,6 +21,9 @@ export class ModalConfig<D = any> {
   replaceTopModal?: boolean;
 }
 
+/**
+ * @deprecated Use the Component Library's `DialogService` instead.
```
Comment (JaredSnider-Bitwarden) [inline libs/angular/src/services/modal.service.ts:25]: Minor question (non-blocking): should we include a target date + tech debt item for cleaning up and removing this service?

**Review by JaredSnider-Bitwarden — APPROVED** [review]
LGTM!

Only a single non blocking question.

**Review by gbubemismith — APPROVED** [review]
LGTM!

**Review by cyprain-okeke — APPROVED** [review]
LGTM!

**Pull request merged**